tools/opensnoop.py

```diff
@@ -235,8 +237,12 @@
 # initialize BPF
 b = BPF(text=bpf_text)
 if not is_support_kfunc:
-    b.attach_kprobe(event="do_sys_open", fn_name="trace_entry")
-    b.attach_kretprobe(event="do_sys_open", fn_name="trace_return")
+    if parse_version(platform.uname().release.split('-')[0]) > parse_version('5.6.0'):
+        entry = "do_sys_openat2"
```

Comment:
This change is tricky. 

First, we cannot do kernel version comparison in tools since there are kernels backported or forward-ported with features and the version checking does not make sense to these kernels. In bcc, we check /proc/kallsyms for a particular global variable to indicate whether a particular feature exists or not.

Second, `do_sys_openat2` is not a safe alternative either. Depending on the compiler, `do_sys_openat2` may be inlined into `do_sys_open`. The safest way is to trace the syscall (__x64_sys_open, __x64_sys_openat, etc.) itself. You can check `execsnoop.py` for an example.

BTW, do_sys_open2 and do_sys_open has different signatures. So this change won't work for kfunc's.

Reply:
Thanks for explanation.

For kprobe part, how about the following one?
https://github.com/iovisor/bcc/pull/3012/commits/5d9963e552a6888e2d39e770597d3cc6c580c7e1

For kfunc part,
1) I was trying to catch __x64_sys_openat but met the following error.

bpf_text_kfunc= """
KRETFUNC_PROBE(__x64_sys_openat, int dfd, const char __user *filename, int flags, int mode, int ret)
...

bpf: Failed to load program: Permission denied
arg#0 type is not a struct
Unrecognized arg#0 type PTR
; KRETFUNC_PROBE(__x64_sys_openat, int dfd, const char __user *filename, int flags, int mode, int ret)
0: (bf) r6 = r1
; KRETFUNC_PROBE(__x64_sys_openat, int dfd, const char __user *filename, int flags, int mode, int ret)
1: (79) r1 = *(u64 *)(r6 +32)
func '__x64_sys_openat' doesn't have 5-th argument
invalid bpf_context access off=32 size=8
processed 2 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0

Traceback (most recent call last):
  File "/usr/share/bcc/tools/opensnoop", line 282, in <module>
    b = BPF(text=bpf_text, debug=4)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 366, in __init__
    self._trace_autoload()
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 1239, in _trace_autoload
    self.attach_kretfunc(fn_name=func_name)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 951, in attach_kretfunc
    fn = self.load_func(fn_name, BPF.TRACING)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 405, in load_func
    raise Exception("Failed to load BPF program %s: %s" %
Exception: Failed to load BPF program b'kretfunc____x64_sys_openat': Permission denied



2) And then I tried to do_sys_openat2, but it seems it's in /proc/kallsyms but not in /sys/kernel/btf/vmlinux

$ sudo /usr/share/bcc/tools/opensnoop
traceback (most recent call last):
  File "/usr/share/bcc/tools/opensnoop", line 283, in <module>
    b = BPF(text=bpf_text, debug=4)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 366, in __init__
    self._trace_autoload()
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 1239, in _trace_autoload
    self.attach_kretfunc(fn_name=func_name)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 951, in attach_kretfunc
    fn = self.load_func(fn_name, BPF.TRACING)
  File "/usr/lib/python3.8/site-packages/bcc/__init__.py", line 405, in load_func
    raise Exception("Failed to load BPF program %s: %s" %
Exception: Failed to load BPF program b'kretfunc__do_sys_openat2': Invalid argument

$ strings /sys/kernel/btf/vmlinux |grep sys_open
__ia32_compat_sys_open
__ia32_compat_sys_open_by_handle_at
__ia32_compat_sys_openat
__ia32_sys_open
__ia32_sys_open_by_handle_at
__ia32_sys_open_tree
__ia32_sys_openat
__ia32_sys_openat2
__x64_sys_open
__x64_sys_open_by_handle_at
__x64_sys_open_tree
__x64_sys_openat
__x64_sys_openat2
do_sys_open

$ cat /proc/kallsyms |grep sys_open
0000000000000000 t do_sys_openat2
0000000000000000 T __ia32_sys_openat2
0000000000000000 T __x64_sys_openat2
0000000000000000 T __x64_sys_open
0000000000000000 T __ia32_compat_sys_open
0000000000000000 T __x64_sys_openat
0000000000000000 T __ia32_sys_open
0000000000000000 T __ia32_compat_sys_openat
0000000000000000 T __ia32_sys_openat
0000000000000000 T do_sys_open
0000000000000000 T __ia32_sys_open_tree
0000000000000000 T __x64_sys_open_tree
0000000000000000 T __x64_sys_open_by_handle_at
0000000000000000 T __ia32_sys_open_by_handle_at
0000000000000000 T __ia32_compat_sys_open_by_handle_at
0000000000000000 t proc_sys_open
0000000000000000 d sys_open_test
0000000000000000 d _eil_addr___ia32_compat_sys_openat
0000000000000000 d _eil_addr___ia32_compat_sys_open
0000000000000000 d _eil_addr___ia32_sys_openat2
0000000000000000 d _eil_addr___x64_sys_openat2
0000000000000000 d _eil_addr___ia32_sys_openat
0000000000000000 d _eil_addr___x64_sys_openat
0000000000000000 d _eil_addr___ia32_sys_open
0000000000000000 d _eil_addr___x64_sys_open
0000000000000000 d _eil_addr___ia32_sys_open_tree
0000000000000000 d _eil_addr___x64_sys_open_tree
0000000000000000 d _eil_addr___ia32_compat_sys_open_by_handle_at
0000000000000000 d _eil_addr___ia32_sys_open_by_handle_at
0000000000000000 d _eil_addr___x64_sys_open_by_handle_at